File: DevExpress VCL/ExpressBars 6/Demos/CBuilder/EBarMDIDemo/EBarMDIDemoChild.cpp
//---------------------------------------------------------------------------
#include <vcl.h>
#pragma hdrstop

#include "EBarMDIDemoChild.h"
#include <stdio.h>
#include "EBarMDIDemoMain.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TEBarMDIDemoChildForm *EBarMDIDemoChildForm;
//---------------------------------------------------------------------------
__fastcall TEBarMDIDemoChildForm::TEBarMDIDemoChildForm(TComponent* Owner)
        : TForm(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TEBarMDIDemoChildForm::SetFileName(AnsiString Value)
{
  int I;

  FFileName = Value;
  Caption = FFileName;
  I = EBarMDIDemoMainForm->dxBarListWindows->Items->IndexOfObject(this);
  if ((0 <= I) && (I < EBarMDIDemoMainForm->dxBarListWindows->Items->Count))
    EBarMDIDemoMainForm->dxBarListWindows->Items->Strings[I] = FFileName;
}
//---------------------------------------------------------------------------

void __fastcall TEBarMDIDemoChildForm::FormCreate(TObject *Sender)
{
  EBarMDIDemoMainForm->CreatedMDICount++;
  Caption = "Document" + IntToStr(EBarMDIDemoMainForm->CreatedMDICount);
  Editor->OnChange = EBarMDIDemoMainForm->EditorChange;
  Editor->OnSelectionChange = EBarMDIDemoMainForm->EditorSelectionChange;
  EBarMDIDemoMainForm->dxBarListWindows->Items->AddObject(Caption, this);
  Editor->Modified;

  if (EBarMDIDemoMainForm->MDIChildCount == 1) EBarMDIDemoMainForm->ShowItems(True);
}
//---------------------------------------------------------------------------

void __fastcall TEBarMDIDemoChildForm::FormClose(TObject *Sender,
      TCloseAction &Action)
{
  if (EBarMDIDemoMainForm->FileName != "")
    EBarMDIDemoMainForm->dxBarMRUFiles->AddItem(EBarMDIDemoMainForm->FileName, NULL);

  Action = caFree;
}
//---------------------------------------------------------------------------

void __fastcall TEBarMDIDemoChildForm::FormActivate(TObject *Sender)
{
  Editor->OnChange(Editor);
}
//---------------------------------------------------------------------------

void __fastcall TEBarMDIDemoChildForm::FormDestroy(TObject *Sender)
{
  EBarMDIDemoMainForm->dxBarListWindows->Items->Delete(EBarMDIDemoMainForm->dxBarListWindows->Items->IndexOfObject(this));

  if (EBarMDIDemoMainForm->MDIChildCount == 1) EBarMDIDemoMainForm->ShowItems(False);
}
//---------------------------------------------------------------------------

void __fastcall TEBarMDIDemoChildForm::FormCloseQuery(TObject *Sender, bool &CanClose)
{
  if (Editor->Modified)
  {
	int Res = MessageBox(0,
	  dxStringToAnsiString(Format("Do you want to save the changes you made to ""%s""?", ARRAYOFCONST((Caption)))).c_str(),
	  dxStringToAnsiString(Application->Title).c_str(), MB_ICONQUESTION | MB_YESNOCANCEL);
	switch (Res)
    {
      case ID_YES: CanClose = EBarMDIDemoMainForm->SaveFile(False); break;
      case ID_NO: CanClose = True; break;
      case ID_CANCEL: CanClose = False; break;
    }
  }
}
//---------------------------------------------------------------------------

void __fastcall TEBarMDIDemoChildForm::EditorMouseDown(TObject *Sender,
      TMouseButton Button, TShiftState Shift, int X, int Y)
{
  if (Button == mbRight)
    EBarMDIDemoMainForm->dxBarPopupMenu->PopupFromCursorPos();
}
//---------------------------------------------------------------------------

File: DevExpress VCL/ExpressBars 6/Demos/Delphi/EBarMDIDemo/EBarMDIDemoChild.pas
unit EBarMDIDemoChild;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs,
  StdCtrls, ComCtrls;

type
  TEBarMDIDemoChildForm = class(TForm)
    Editor: TRichEdit;
    procedure FormCreate(Sender: TObject);
    procedure FormClose(Sender: TObject; var Action: TCloseAction);
    procedure FormActivate(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure FormCloseQuery(Sender: TObject; var CanClose: Boolean);
    procedure EditorMouseDown(Sender: TObject; Button: TMouseButton;
      Shift: TShiftState; X, Y: Integer);
  private
    FFileName: string;
    procedure SetFileName(Value: string);
  public
    property FileName: string read FFileName write SetFileName;
  end;

implementation

uses EBarMDIDemoMain;

{$R *.DFM}

procedure TEBarMDIDemoChildForm.SetFileName(Value: string);
var
  I: Integer;
begin
  FFileName := Value;
  Caption := FFileName;
  with EBarMDIDemoMainForm.dxBarListWindows do
  begin
    I := Items.IndexOfObject(Self);
    if (0 <= I) and (I < Items.Count) then Items[I] := FFileName;
  end;
end;

procedure TEBarMDIDemoChildForm.FormCreate(Sender: TObject);
begin
  Inc(EBarMDIDemoMainForm.CreatedMDICount);
  Caption := 'Document' + IntToStr(EBarMDIDemoMainForm.CreatedMDICount);
  Editor.OnChange := EBarMDIDemoMainForm.EditorChange;
  Editor.OnSelectionChange := EBarMDIDemoMainForm.EditorSelectionChange;
  EBarMDIDemoMainForm.dxBarListWindows.Items.AddObject(Caption, Self);
  Editor.Modified;

  if EBarMDIDemoMainForm.MDIChildCount = 1 then EBarMDIDemoMainForm.ShowItems(True);
end;

procedure TEBarMDIDemoChildForm.FormClose(Sender: TObject; var Action: TCloseAction);
begin
  if EBarMDIDemoMainForm.FileName <> '' then
    EBarMDIDemoMainForm.dxBarMRUFiles.AddItem(EBarMDIDemoMainForm.FileName, nil);
  Action := caFree;
end;

procedure TEBarMDIDemoChildForm.FormActivate(Sender: TObject);
begin
  Editor.OnChange(Editor);
end;

procedure TEBarMDIDemoChildForm.FormDestroy(Sender: TObject);
begin
  with EBarMDIDemoMainForm.dxBarListWindows.Items do
    Delete(IndexOfObject(Self));

  if EBarMDIDemoMainForm.MDIChildCount = 1 then EBarMDIDemoMainForm.ShowItems(False);
end;

procedure TEBarMDIDemoChildForm.FormCloseQuery(Sender: TObject;
  var CanClose: Boolean);
var
  Res: Integer;
begin
  if Editor.Modified then
  begin
    Res := Application.MessageBox(
      PChar(Format('Do you want to save the changes you made to "%s"?', [Caption])),
      PChar(Application.Title), MB_ICONQUESTION or MB_YESNOCANCEL);
    case Res of
      ID_YES: CanClose := EBarMDIDemoMainForm.SaveFile(False);
      ID_NO: CanClose := True;
      ID_CANCEL: CanClose := False;
    end;
  end;
end;

procedure TEBarMDIDemoChildForm.EditorMouseDown(Sender: TObject; Button: TMouseButton;
  Shift: TShiftState; X, Y: Integer);
begin
  if Button = mbRight then
    EBarMDIDemoMainForm.dxBarPopupMenu.PopupFromCursorPos;
end;

end.
